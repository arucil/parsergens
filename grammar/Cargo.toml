[package]
name = "grammar"
version = "0.1.0"
authors = ["amlo <xplzjwz@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lalrpop-util = "0.19.0"
bit-set = "0.5.2"
indexmap = "1.5.1"
bimap = "0.5.2"
either = "1.6.0"
codespan-reporting = "0.9.5"
bitvec = "0.18.1"

[build-dependencies]
lalrpop = "0.19.0"

[dev-dependencies]
insta = "0.16.1"
pretty_assertions = "0.6.1"
