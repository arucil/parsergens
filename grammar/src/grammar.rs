use std::ops::Range;
use std::borrow::Borrow;
use super::lexer::{Lexer, TokenId};
use crate::{Set, BiMap, Map};

pub use lower::*;

pub mod lower;

#[derive(Debug)]
pub struct Grammar {
  /// productions of the same nonterminals are consecutive.
  pub rules: Vec<Rule>,
  /// starting non-terminals
  pub start_nts: Set<NonterminalId>,
  /// non-terminals
  pub nts: BiMap<NonterminalId, String>,
  pub nt_prods: Map<NonterminalId, Range<usize>>,
  pub lexer: Lexer,
}

#[derive(Debug)]
pub struct Rule {
  pub nt: NonterminalId,
  pub items: Vec<Item>,
}

#[derive(Debug)]
pub enum Item {
  Nonterminal(NonterminalId),
  Token(TokenId),
  Optional(Vec<Item>),
  Many(Vec<Item>),
  Many1(Vec<Item>),
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
pub struct NonterminalId(u32);

impl NonterminalId {
  pub fn id(&self) -> u32 {
    self.0
  }
}

impl Borrow<u32> for NonterminalId {
  fn borrow(&self) -> &u32 {
    &self.0
  }
}

#[derive(Default)]
pub struct NonterminalIdGen(u32);

impl NonterminalIdGen {
  pub fn gen(&mut self) -> NonterminalId {
    let i = self.0;
    self.0 += 1;
    NonterminalId(i)
  }

  pub fn from(start: u32) -> Self {
    Self(start + 1)
  }
}

impl Grammar {
  pub fn lower(self) -> LoweredGrammar {
    lower::lower(self)
  }
}