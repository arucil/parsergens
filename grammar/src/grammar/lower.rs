use std::ops::Range;
use crate::{
  Grammar, NonterminalIdGen, Item, Lexer, TokenId, NonterminalId
};
use crate::{Set, BiMap, Map};

#[derive(Debug)]
pub struct LoweredGrammar {
  pub prods: Vec<Production>,
  pub start_nts: Set<NonterminalId>,
  pub nts: BiMap<NonterminalId, String>,
  pub nt_metas: Map<NonterminalId, LoweredNonterminalMetadata>,
  pub lexer: Option<Lexer>,
  pub tokens: BiMap<TokenId, String>,
  pub user_code: Vec<String>,
}

#[derive(Debug)]
pub struct Production {
  pub nt: NonterminalId,
  pub kind: ProductionKind,
  pub symbols: Vec<Symbol>,
  pub action: Option<String>,
}

#[derive(Debug)]
pub struct LoweredNonterminalMetadata {
  pub range: Range<usize>,
  pub ty: Option<String>,
  pub kind: NonterminalKind,
}

#[derive(Debug, Clone, Copy)]
pub enum NonterminalKind {
  User,
  Repetition,
  Optional,
}

#[derive(Debug, Clone, Copy)]
pub enum ProductionKind {
  Ordinary,
  RepetitionFirst,
  RepetitionRest,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Symbol {
  Nonterminal(NonterminalId),
  Token(TokenId),
}

pub(super) fn lower(grammar: Grammar) -> LoweredGrammar {
  let max_nt_id = grammar.nts
    .left_values()
    .map(|x| x.id())
    .max()
    .unwrap();
  let mut nt_id_gen = NonterminalIdGen::from(max_nt_id);

  let mut lowered = LoweredGrammar {
    prods: vec![],
    start_nts: grammar.start_nts,
    nts: grammar.nts,
    nt_metas: Map::new(),
    lexer: grammar.lexer,
    tokens: grammar.tokens,
    user_code: grammar.user_code,
  };

  for (nt, meta) in grammar.nt_metas {
    let mut rules = vec![];

    for rule in &grammar.rules[meta.range] {
      rules.push((
        lower_items(&rule.items, &mut lowered, &mut nt_id_gen),
        rule.action.clone()
      ));
    }

    let start = lowered.prods.len();

    for (symbols, action) in rules {
      lowered.prods.push(Production {
        nt,
        kind: ProductionKind::Ordinary,
        symbols,
        action,
      });
    }

    let end = lowered.prods.len();

    let meta = LoweredNonterminalMetadata {
      range: start..end,
      ty: meta.ty,
      kind: NonterminalKind::User,
    };

    lowered.nt_metas.insert(nt, meta);
  }

  lowered
}

fn lower_items(
  items: &[Item],
  lowered: &mut LoweredGrammar,
  nt_id_gen: &mut NonterminalIdGen,
) -> Vec<Symbol> {
  items.iter().map(|item| {
    match item {
      Item::Nonterminal(nt) => Symbol::Nonterminal(*nt),
      Item::Token(token) => Symbol::Token(*token),
      Item::Optional(items) => {
        let symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '?');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::Ordinary,
          symbols: vec![],
          action: None,
        });
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::Ordinary,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Optional,
        });

        Symbol::Nonterminal(nt)
      }
      Item::Many(items) => {
        let mut symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '*');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionFirst,
          symbols: vec![],
          action: None,
        });

        symbols.insert(0, Symbol::Nonterminal(nt));
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionRest,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Repetition,
        });

        Symbol::Nonterminal(nt)
      }
      Item::Many1(items) => {
        let mut symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '+');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionFirst,
          symbols: symbols.clone(),
          action: None,
        });

        symbols.insert(0, Symbol::Nonterminal(nt));
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionRest,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Repetition,
        });

        Symbol::Nonterminal(nt)
      }
    }
  }).collect()
}

fn make_production_name(
  symbols: &[Symbol],
  lowered: &LoweredGrammar,
  suffix: char
) -> String {
  let mut buf = "(".to_owned();
  let mut space = false;

  for symbol in symbols {
    if space {
      buf.push(' ');
    }

    match symbol {
      Symbol::Nonterminal(nt) => {
        buf.push_str(lowered.nts.get_by_left(nt).unwrap());
      }
      Symbol::Token(token) => {
        buf.push_str(lowered.tokens.get_by_left(token).unwrap());
      }
    }

    space = true;
  }

  buf.push(')');
  buf.push(suffix);

  buf
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn lower_repetition() {
    let grammar = crate::build(r#"
%token A "a"

%start top

top = A (atom (A)? )* A
  | (A atom)+

atom = A
  | ()
    "#).unwrap();

    let lowered = lower(grammar);

    assert_debug_snapshot!(lowered);
  }
}