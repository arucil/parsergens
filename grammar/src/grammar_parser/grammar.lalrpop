use lalrpop_util::ParseError;
use super::ast::*;
use super::lex;
use super::UserParseError;
use super::regex;

#[LALR]
grammar<'input>(input: &'input str);

pub Document: Grammar = (Separator <decl>)*
  => {
    <>.into_iter()
      .filter_map(|x| x)
      .collect()
  };

decl: Option<Spanned<Decl>> = {
  <l:@L> <decl:rule_decl> <r:@L> => Some(Spanned((l, r), Decl::Rule(decl))),
  <l:@L> <decl:start_decl> <r:@L> => Some(Spanned((l, r), Decl::Start(decl))),
  <l:@L> <decl:token_decl> <r:@L> => Some(Spanned((l, r), Decl::Token(decl))),
  <l:@L> <decl:ext_token_decl> <r:@L> => Some(Spanned((l, r), Decl::ExtToken(decl))),
  <l:@L> <decl:skip_decl> <r:@L> => Some(Spanned((l, r), Decl::Skip(decl))),
  <l:@L> <decl:user_decl> <r:@L> => Some(Spanned((l, r), Decl::User(decl))),
  <l:@L> <decl:assoc_decl> <r:@L> => Some(Spanned((l, r), Decl::Assoc(decl))),
  <l:@L> <decl:state_decl> <r:@L> => Some(Spanned((l, r), Decl::State(decl))),
};

rule_decl: RuleDecl =
  <name:ident> <ty:(Colon <code_block>)?> Assign <alt1:production> <mut alts:(Or <production>)*>
  =>? {
    alts.insert(0, alt1);
    if alts.iter().any(|alt| alt.1.action.is_some()) && ty.is_none() {
      Err(ParseError::User {
        error: UserParseError::MissingType(name.0),
      })
    } else {
      Ok(RuleDecl {
        name,
        ty,
        alts,
      })
    }
  }; 

production: Spanned<RuleAlt> = {
  <l:@L> <terms:term+> <prec:(Percent Prec <ident>)?> <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Terms(terms),
        prec,
        action,
      }
    ),
  <l:@L> epsilon <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Epsilon,
        prec: None,
        action,
      }
    ),
};

epsilon: () = LParen RParen;

term: Term = {
  ident => Term::Symbol(<>),
  LParen <terms:term+> RParen QuestionMark => Term::Optional(terms),
  LParen <terms:term+> RParen Asterisk => Term::Many(terms),
  LParen <terms:term+> RParen Plus => Term::Many1(terms),
}

start_decl: StartDecl = Percent Start <name:ident>
  => {
    StartDecl {
      name,
    }
  };

token_decl: TokenDecl = Percent Token <name:ident> <pattern:token_pattern>
  => {
    TokenDecl {
      name,
      pattern,
    }
  };

ext_token_decl: ExternalTokenDecl = Percent Token <name:ident>
  => {
    ExternalTokenDecl {
      name,
    }
  };

skip_decl: SkipDecl = Percent Skip <pattern:token_pattern>
  => {
    SkipDecl {
      pattern,
    }
  };

user_decl: UserDecl = Percent User <code:code_block>
  => {
    UserDecl {
      code,
    }
  };

assoc_decl: AssocDecl = Percent <assoc:assoc> <names:ident+>
  => {
    AssocDecl {
      names,
      assoc,
    }
  };

assoc: Spanned<Assoc> = {
  <l:@L> LeftAssoc <r:@L> => Spanned((l, r), Assoc::LeftAssoc),
  <l:@L> RightAssoc <r:@L> => Spanned((l, r), Assoc::RightAssoc),
  <l:@L> NonAssoc <r:@L> => Spanned((l, r), Assoc::NonAssoc),
}

state_decl: StateDecl = Percent State <state:code_block>
  => {
    StateDecl {
      state,
    }
  };

token_pattern: TokenPattern = {
  <regex:regex> =>? regex::parse_regex(
    &regex.1[1..regex.1.len() - 1],
    regex.0 .0 + 1
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::Regex,
      source: regex,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error),
    }),
  <str:string> =>? regex::parse_string_literal(
    &str.1[1..str.1.len() - 1], str.0 .0
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::String,
      source: str,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error)
    }),
}

spanned<T>: Spanned<String> = <l:@L> <x:T> <r:@L> => Spanned((l, r), x.text.to_owned());

ident = spanned<Ident>;
string = spanned<String>;
regex = spanned<Regex>;
code_block: Spanned<String> = <spanned<CodeBlock>> =>
  Spanned(<>.0, <>.1.trim_start_matches('{').trim_end_matches('}').trim().to_owned());

extern {
  type Location = usize;
  type Error = UserParseError;

  enum lex::Token<'input> {
    Start => lex::Token { kind: lex::TokenKind::Start, .. },
    Token => lex::Token { kind: lex::TokenKind::Token, .. },
    Skip => lex::Token { kind: lex::TokenKind::Skip, .. },
    User => lex::Token { kind: lex::TokenKind::User, .. },
    LeftAssoc => lex::Token { kind: lex::TokenKind::LeftAssoc, .. },
    RightAssoc => lex::Token { kind: lex::TokenKind::RightAssoc, .. },
    NonAssoc => lex::Token { kind: lex::TokenKind::NonAssoc, .. },
    Prec => lex::Token { kind: lex::TokenKind::Prec, .. },
    State => lex::Token { kind: lex::TokenKind::State, .. },

    Ident => lex::Token { kind: lex::TokenKind::Ident, .. },
    Regex => lex::Token { kind: lex::TokenKind::Regex, .. },
    String => lex::Token { kind: lex::TokenKind::String, .. },
    CodeBlock => lex::Token { kind: lex::TokenKind::CodeBlock, .. },

    Percent => lex::Token { kind: lex::TokenKind::Percent, .. },
    Assign => lex::Token { kind: lex::TokenKind::Assign, .. },
    Or => lex::Token { kind: lex::TokenKind::Or, .. },
    LParen => lex::Token { kind: lex::TokenKind::LParen, .. },
    RParen => lex::Token { kind: lex::TokenKind::RParen, .. },
    Comma => lex::Token { kind: lex::TokenKind::Comma, .. },
    Colon => lex::Token { kind: lex::TokenKind::Colon, .. },
    Asterisk => lex::Token { kind: lex::TokenKind::Asterisk, .. },
    QuestionMark => lex::Token { kind: lex::TokenKind::QuestionMark, .. },
    Plus => lex::Token { kind: lex::TokenKind::Plus, .. },
    Arrow => lex::Token { kind: lex::TokenKind::Arrow, .. },

    Enter => lex::Token { kind: lex::TokenKind::Enter, .. },
    Leave => lex::Token { kind: lex::TokenKind::Leave, .. },
    Separator => lex::Token { kind: lex::TokenKind::Separator, .. },
  }
}