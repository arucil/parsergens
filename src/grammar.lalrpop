use super::ast::*;
use super::lex;

#[LALR]
grammar<'input>(input: &'input str);

pub document: Grammar = <decl*>
  => {
    <>.into_iter()
      .filter_map(|x| x)
      .collect()
  };

decl: Option<Spanned<Decl>> = {
  <l:@L> <decl:rule_decl> <r:@L> => Some(Spanned(l..r, Decl::Rule(decl))),
  <l:@L> <decl:start_decl> <r:@L> => Some(Spanned(l..r, Decl::Start(decl))),
  <l:@L> <decl:token_decl> <r:@L> => Some(Spanned(l..r, Decl::Token(decl))),
  Newline => None,
};

rule_decl: RuleDecl =
  <name:ident> Assign <alt1:production> Newline <mut alts:(Indent Or <production> Newline)*>
  => {
    alts.insert(0, alt1);
    RuleDecl {
      name,
      alts,
    }
  }; 

production: Spanned<RuleAlt> = {
  <l:@L> <terms:ident*> <r:@L> => Spanned(l..r, RuleAlt::Terms(terms)),
  <l:@L> Epsilon <r:@L> => Spanned(l..r, RuleAlt::Epsilon),
};

start_decl: StartDecl = Percent Start <name:ident> Newline
  => {
    StartDecl {
      name,
    }
  };

token_decl: TokenDecl = Percent Token <name:ident> <regex:regex>
  => {
    TokenDecl {
      name,
      regex,
    }
  };

spanned<T>: Spanned<String> = <l:@L> <x:T> <r:@L> => Spanned(l..r, x.text.to_owned());

ident = spanned<Ident>;
regex = spanned<Regex>;

extern {
  type Location = usize;
  type Error = lex::LexError;

  enum lex::Token<'input> {
    Ident => lex::Token { kind: lex::TokenKind::Ident, .. },
    Start => lex::Token { kind: lex::TokenKind::Start, .. },
    Token => lex::Token { kind: lex::TokenKind::Token, .. },
    Regex => lex::Token { kind: lex::TokenKind::Regex, .. },

    Percent => lex::Token { kind: lex::TokenKind::Percent, .. },
    Assign => lex::Token { kind: lex::TokenKind::Assign, .. },
    Or => lex::Token { kind: lex::TokenKind::Or, .. },
    Epsilon => lex::Token { kind: lex::TokenKind::Epsilon, .. },
    Comma => lex::Token { kind: lex::TokenKind::Comma, .. },

    Indent => lex::Token { kind: lex::TokenKind::Indent, .. },
    Newline => lex::Token { kind: lex::TokenKind::Newline, .. },
  }
}