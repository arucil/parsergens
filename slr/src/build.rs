use std::collections::VecDeque;
use bit_set::BitSet;
use grammar::{GrammarError, TokenId, NonterminalId, NonterminalIdGen};
use crate::{SlrParser, BiMap, Map};
use lower::{LoweredGrammar, Symbol};
use ffn::Ffn;

mod ffn;
mod lower;
mod augment;

#[derive(Debug)]
pub enum Error {
  GrammarError(GrammarError),
  ShiftReduceConflict,
  ReduceReduceConflict,
}

pub(crate) fn build(input: &str) -> Result<SlrParser, Error> {
  let grammar = grammar::build(input).map_err(Error::GrammarError)?;
  let grammar = lower::lower(grammar);
  let (grammar, eof_token) = augment::augment(grammar);
  let ffn = ffn::compute(&grammar);

  let mut builder = Builder::new(&grammar, eof_token, ffn);

  builder.build()?;

  let prods = grammar.prods.iter().map(|prod| {
    (prod.symbols.len(), prod.nt.id())
  }).collect();

  let nt_names = (0..grammar.nts.len()).scan(
    NonterminalIdGen::default(),
    |gen, _| Some(gen.gen()))
    .map(|nt| grammar.nts.get_by_left(&nt).unwrap().clone())
    .collect();

  Ok(SlrParser {
    action: builder.action,
    goto: builder.goto,
    prods,
    nt_names,
    start: builder.start,
    eof_index: eof_token.id() as usize,
    lexer: grammar.lexer
  })
}

struct Builder<'a> {
  grammar: &'a LoweredGrammar,
  ffn: Ffn,
  states: Map<BitSet, u32>,
  items: BiMap<(usize, usize), usize>,
  action: Vec<Vec<i32>>,
  action_row_len: usize,
  goto: Vec<Vec<u32>>,
  goto_row_len: usize,
  start: Map<String, u32>,
  eof_token: TokenId,
}

impl<'a> Builder<'a> {
  fn new(grammar: &'a LoweredGrammar, eof_token: TokenId, ffn: Ffn) -> Self {
    Builder {
      grammar: &grammar,
      ffn,
      states: Map::new(),
      items: BiMap::new(),
      action: vec![],
      action_row_len: eof_token.id() as usize + 1,
      goto: vec![],
      goto_row_len: grammar.nts.len(),
      start: Map::new(),
      eof_token,
    }
  }

  fn build(&mut self) -> Result<(), Error> {
    for &start_nt in &self.grammar.start_nts {
      let start_state = self.start(start_nt)?;
      let nt_name = self.grammar.nts.get_by_left(&start_nt).unwrap().clone();
      self.start.insert(nt_name, start_state);
    }

    Ok(())
  }

  fn start(&mut self, nt: NonterminalId) -> Result<u32, Error> {
    let start_prod = self.grammar.nt_prods[&nt].start;
    let start_item = self.item(start_prod, 0);
    let start_state_set = self.closure({
      let mut set = BitSet::new();
      set.insert(start_item);
      set
    });
    let start_state = self.state(&start_state_set);

    let mut queue = VecDeque::new();
    queue.push_back(start_state_set);

    while let Some(state) = queue.pop_front() {
      let from_state = self.state(&state) as usize;
      let mut to_states = Map::<Symbol, BitSet>::new();

      for item in state.iter() {
        let &(prod_ix, prod_dot) = self.items.get_by_right(&item).unwrap();
        let symbols = &self.grammar.prods[prod_ix].symbols;
        if prod_dot == symbols.len() {
          self.reduce(from_state, prod_ix)?;
          continue;
        }

        if prod_dot == 1 && prod_ix == start_prod {
          self.accept(from_state);
          continue;
        }

        let new_item = self.item(prod_ix, prod_dot + 1);
        to_states.entry(symbols[prod_dot].clone()).or_default().insert(new_item);
      }

      for (sym, to_state) in to_states {
        let to_state = self.closure(to_state);
        let is_new_state = !self.states.contains_key(&to_state);
        let to_state_ix = self.state(&to_state);

        match sym {
          Symbol::Token(token) => {
            let old = self.action[from_state][token.id() as usize];
            if old < 0 {
              return Err(Error::ShiftReduceConflict);
            }
            assert!(old == 0);

            self.action[from_state][token.id() as usize] = (to_state_ix + 1) as i32;
          }
          Symbol::Nonterminal(nt) => {
            assert!(self.goto[from_state][nt.id() as usize] == 0);

            self.goto[from_state][nt.id() as usize] = to_state_ix + 1;
          }
        }

        if is_new_state {
          queue.push_back(to_state);
        }
      }
    }

    Ok(start_state)
  }

  fn closure(&mut self, initial: BitSet) -> BitSet {
    let mut result = initial.clone();
    let mut last = initial;

    loop {
      let mut new = BitSet::new();
      for i in last.iter() {
        let &(prod_ix, prod_dot) = self.items.get_by_right(&i).unwrap();
        let symbols = &self.grammar.prods[prod_ix].symbols;
        if prod_dot < symbols.len() {
          match &symbols[prod_dot] {
            Symbol::Token(_) => {}
            Symbol::Nonterminal(nt) => {
              for prod_ix in self.grammar.nt_prods[nt].clone() {
                let item = self.item(prod_ix, 0);
                if !result.contains(item) {
                  new.insert(item);
                }
              }

              /*
              if self.ffn.nullable.contains(nt.id() as usize) {
                let item = self.item(prod_ix, prod_dot + 1);
                if !result.contains(item) {
                  new.insert(item);
                }
              }
              */
            }
          }
        }
      }

      if new.is_empty() {
        break;
      }

      result.union_with(&new);

      last = new;
    }

    result
  }

  fn item(&mut self, prod_ix: usize, prod_dot: usize) -> usize {
    if let Some(&item) = self.items.get_by_left(&(prod_ix, prod_dot)) {
      item
    } else {
      let len = self.items.len();
      self.items.insert((prod_ix, prod_dot), len);
      len
    }
  }

  fn state(&mut self, set: &BitSet) -> u32 {
    if let Some(&state) = self.states.get(set) {
      state
    } else {
      let len = self.states.len() as u32;
      self.states.insert(set.clone(), len);

      self.action.push(vec![0; self.action_row_len]);
      self.goto.push(vec![0; self.goto_row_len]);

      len
    }
  }

  fn reduce(&mut self, from_state: usize, prod_ix: usize) -> Result<(), Error> {
    let nt = self.grammar.prods[prod_ix].nt;

    for token in self.ffn.follow[&nt].iter() {
      let old = self.action[from_state][token];
      if old < 0 {
        //return Err(Error::ReduceReduceConflict);
        println!(">>>>>>>>>>>>>>> reduce-reduce");
      }
      //assert!(old == 0);

      self.action[from_state][token] = !(prod_ix as i32);
    }

    Ok(())
  }

  fn accept(&mut self, from_state: usize) {
    self.action[from_state][self.eof_token.id() as usize] = std::i32::MIN;
  }
}

#[cfg(test)]
use std::fmt::{self, Write};

#[cfg(test)]
impl<'a> Builder<'a> {
  fn states(self) -> String {
    let mut buf = String::new();
    self.fmt_states(&mut buf).unwrap();
    buf
  }

  fn fmt_states(self, fmt: &mut impl Write) -> fmt::Result {
    let mut states = self.states.into_iter().collect::<Vec<_>>();
    states.sort_by_key(|(_, x)| *x);

    for (state_set, state) in states {
      write!(fmt, "State {}", state)?;
      if self.start.values().find(|&&x| x == state).is_some() {
        write!(fmt, " (start)")?;
      }
      writeln!(fmt)?;

      for item in state_set.iter() {
        let &(prod_ix, prod_dot) = self.items.get_by_right(&item).unwrap();
        let nt = self.grammar.prods[prod_ix].nt;
        let symbols = &self.grammar.prods[prod_ix].symbols;

        write!(fmt, "{} ->", self.grammar.nts.get_by_left(&nt).unwrap())?;

        for (i, sym) in symbols.iter().enumerate() {
          if i == prod_dot {
            write!(fmt, " .")?;
          }

          match sym {
            Symbol::Token(token) => {
              let name = self.grammar.lexer.tokens.get_by_left(token).map(|s|s.as_str()).unwrap_or("$");
              write!(fmt, " {}", name)?;
            }
            Symbol::Nonterminal(nt) => {
              let name = self.grammar.nts.get_by_left(nt).unwrap();
              write!(fmt, " {}", name)?;
            }
          }
        } 

        if prod_dot == symbols.len() {
          write!(fmt, " .")?;
        }

        writeln!(fmt)?;
      }

      writeln!(fmt)?;
    }

    Ok(())
  }

  fn action_goto(self) -> Vec<Vec<i32>> {
    self.action.into_iter().zip(self.goto).map(|(mut row1, row2)| {
      row1.extend(row2.into_iter().map(|x| x as i32));
      row1
    }).collect()
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::{assert_debug_snapshot, assert_snapshot};

  fn prepare(input: &str) -> (LoweredGrammar, TokenId, Ffn) {
    let grammar = grammar::build(input).unwrap();
    let grammar = lower::lower(grammar);
    let (grammar, eof_token) = augment::augment(grammar);
    let ffn = ffn::compute(&grammar);

    (grammar, eof_token, ffn)
  }

  static SIMPLE: &str = r#"
%token x "x"
%token plus "+"

%start E

E = T plus E
  | T

T = x
  "#;

  #[test]
  fn simple_states() {
    let (grammar, eof_token, ffn) = prepare(SIMPLE);
    let mut builder = Builder::new(&grammar, eof_token, ffn);

    builder.build().unwrap();

    assert_snapshot!(builder.states());
  }

  #[test]
  fn simple_action_goto() {
    let (grammar, eof_token, ffn) = prepare(SIMPLE);
    let mut builder = Builder::new(&grammar, eof_token, ffn);

    builder.build().unwrap();

    assert_debug_snapshot!(builder.action_goto());
  }

  static EPSILON: &str = r#"
%token plus "+"
%token mult "*"
%token num "1"
%token lparen "("
%token rparen ")"

%start E

E = T E_

E_ = plus T E_
   | ()

T = F T_

T_ = mult F T_
   | ()

F = num
  | lparen E rparen
  "#;

  #[test]
  fn epsilon_states() {
    let (grammar, eof_token, ffn) = prepare(EPSILON);
    let mut builder = Builder::new(&grammar, eof_token, ffn);

    builder.build().unwrap();

    assert_snapshot!(builder.states());
  }

  #[test]
  fn epsilon_action_goto() {
    let (grammar, eof_token, ffn) = prepare(EPSILON);
    let mut builder = Builder::new(&grammar, eof_token, ffn);

    builder.build().unwrap();

    assert_debug_snapshot!(builder.action_goto());
  }
}