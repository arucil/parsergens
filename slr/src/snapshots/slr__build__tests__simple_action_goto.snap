---
source: slr/src/build.rs
expression: builder.action_goto()
---
[
    [
        4,
        0,
        0,
        2,
        3,
        0,
    ],
    [
        0,
        0,
        -2147483648,
        0,
        0,
        0,
    ],
    [
        0,
        5,
        -2,
        0,
        0,
        0,
    ],
    [
        0,
        -3,
        -3,
        0,
        0,
        0,
    ],
    [
        4,
        0,
        0,
        6,
        3,
        0,
    ],
    [
        0,
        0,
        -1,
        0,
        0,
        0,
    ],
]
